types/valuerenderer/valuerenderer_test.go

```diff
@@ -0,0 +1,328 @@
+
+	queryHelper := baseapp.NewQueryServerTestHelper(ctx, app.InterfaceRegistry())
+	banktypes.RegisterQueryServer(queryHelper, app.BankKeeper)
+	queryClient := banktypes.NewQueryClient(queryHelper)
+
+	suite.app = app
+	suite.ctx = ctx
+	suite.queryClient = queryClient
+}
+
+// TODO
+// 1. add some test cases for integer overflow, underflow
+func (suite *valueRendererTestSuite) TestFormatDenomQuerierFunc() {
+
+	metadataRegen := banktypes.Metadata{
+		Name:        "Regen",
+		Symbol:      "REGEN",
+		Description: "The native staking token of the Regen network.",
+		DenomUnits: []*banktypes.DenomUnit{
+			{
+				Denom:    "uregen",
+				Exponent: 0,
+				Aliases:  []string{"microregen"},
+			},
+			{
+				Denom:    "mregen",
+				Exponent: 3,
+				Aliases:  []string{"milliregen"},
+			},
+			{
+				Denom:    "regen",
+				Exponent: 6,
+				Aliases:  []string{"regen"},
+			},
+		},
+		Base:    "uregen",
+		Display: "regen",
+	}
+
+	tt := []struct {
+		name    string
+		coin    types.Coin
+		pretest func()
+		expRes  string
+		expErr  bool
+	}{
+		{
+			"convert 1000000uregen to 1regen",
+			types.NewCoin("uregen", types.NewInt(int64(1000000))),
+			func() {
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, metadataRegen)
+			},
+			"1regen",
+			false,
+		},
+		{
+			"convert 23000mregen to 23regen",
+			types.NewCoin("mregen", types.NewInt(int64(23000))),
+			func() {
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, metadataRegen)
+			},
+			"23regen",
+			false,
+		},
+		{
+			"invalid coin denom",
+			types.NewCoin("atom", types.NewInt(int64(23000))),
+			func() {
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, metadataRegen)
+			},
+			"",
+			true,
+		},
+		{
+			"convert 20000000000uatom to 20000atom, multiple denoms",
+			types.NewCoin("uatom", types.NewInt(int64(20000000000))),
+			func() {
+				metadataAtom := banktypes.Metadata{
+					Name:        "Cosmos Hub Atom",
+					Symbol:      "ATOM",
+					Description: "The native staking token of the Cosmos Hub.",
+					DenomUnits: []*banktypes.DenomUnit{
+						{"uatom", uint32(0), []string{"microatom"}},
+						{"matom", uint32(3), []string{"milliatom"}},
+						{"atom", uint32(6), nil},
+					},
+					Base:    "uatom",
+					Display: "atom",
+				}
+
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, metadataRegen)
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, metadataAtom)
+			},
+			"20,000atom",
+			false,
+		},
+		{
+			"convert 23mregen to 0,023regen",
+			types.NewCoin("mregen", types.NewInt(int64(23))),
+			func() {
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, metadataRegen)
+			},
+			"0.023regen",
+			false,
+		},
+	}
+
+	for _, tc := range tt {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+			tc.pretest()
+
+			c := types.WrapSDKContext(suite.ctx)
+
+			dvr := valuerenderer.NewDefaultValueRenderer(func(c context.Context, baseDenom string) (banktypes.Metadata, error) {
+				req := &banktypes.QueryDenomMetadataRequest{
+					Denom: baseDenom,
```

Comment:
Since `baseDenom` is used in `req` at `queryClient.DenomMetadata`  call , I try to find a proper way to convert `denom` to 'baseDenom`.  [convertToBaseDenom](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/value-renderer/types/valuerenderer/valuerenderer.go#L91) is a temporary solution.